keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,504 @@
+IDL. That marker will be transformed into the OpenAPI `default` field
+and then routed to defaulting functions so that defaulting can be done
+declaratively. This new marker would also be introduced in kubebuilder,
+and we will make the behavior equivalent between built-in types and CRDs
+so that built-in types can be directly used in kubebuilder, and the
+declarative defaults will be automatically carried along.
+
+### Non-Goals
+
+While this proposal is probably paving the way for additional OpenAPI
+based validation, validation is entirely out-of-scope for this project.
+
+This proposal also doesn't intend to remove the existing defaulting
+mechanism since some defaulting still requires non-static values that
+need to be computed and can't be defined declaratively. We could replace
+some defaulting that doesn't require to be done as code though.
+
+## Proposal
+
+It is important to us that the OpenAPI definition of a type should
+result in the same behavior, whether the type is built-in or a CRD.
+
+In order to do that, we need to change how CRDs are defaulted, and
+create a defaulting mechanism for built-in that would behave the same
+way.
+
+### CRD Normalization
+
+Null values in CRDs are currently not defaulted. On the other hand,
+`null` typically means "unspecified" in json/yaml. JSon Merge patch
+interprets a "null" value for a field as a mean to delete it.
+
+On the other hand, while `{field: null}` and `{}` probably have the same
+meaning, they do not "look" the same and are hard to compare for
+clients. They are also currently treated differently by Kubernetes and
+for defaulting purposes.
+
+We propose the creation of a new top-level openapi flag:
+`x-kubernetes-remove-nulls`. This flag will be used as a ratcheting
+mechanism as we migrate to this new semantic throughout all Kubernetes
+types (which would be how all types eventually work). When present and
+set to true, the apiserver will use a new semantics and remove all null
+values from CR objects. When false or missing, the existing and
+deprecated semantics will be kept. Alternatively, this flag could be set
+as a field of the CRD since this process will only apply to CRDs.
+
+Note that we still consider `{}` to be different from `null`, so
+`{empty: {}}` is different from `{}` or from `{empty: null}` (but the
+latter two are considered equal).
+
+Note that this mechanism will also help with immutability.
+
+### Built-in Defaulting
+
+Because built-in types are serialized and deserialized as Go structures,
+there are semantics that are imposed to us. The two main ones are:
+1. We can't differentiate between null, unspecified and zero-value,
+2. non-pointer structs are always present (and hence have to be
+recursively defaulted), they also can't be omitEmpty'd.
+
+The first point is solved by the removed nulls described in the previous
+section (and invalid zero-values can't be present in objects), and the
```

Comment:
> (and invalid zero-values can't be present in objects)

Is this a new requirement?

Reply:
No

Reply:
I'll clarify that, thanks. They are invalid so they shouldn't be specified as actual values :-)
```suggestion
section (and invalid zero-values can't be present in objects since they are invalid), and the
```